Added Hardware PWM Control

Fionnoch opened the pull request:
Added hardware PWM using sysfs based on https://developer.toradex.com/knowledge-base/pwm-linux

It has been created as a class in order to create better tracking of PWM parameters, this is important as PWM has more features than a standard GPIO output.

This was developed and tested on an OrangePi PC Plus and works :smile:  however was not tested on other boards. Likely though it will work on any H3 chip. The only thing to bare in mind is that you need to use PWM0 pin. On the OPi PC+ board this is the middle pin of the Debug TTL UART pins, located between the power and HDMI. The documentation mentions a PWM1 at pin A6 (physical pin 7) but I could not figure out how to access that. I believe the DTC will have to be changed to access it.

The commands that are used to create and control the PWM signal are:
p = GPIO.PWM(PWM_pin, frequency_Hz, Duty_Cycle_Percent) to create the object
p.start_pwm()      To begin the output
p.duty_cycle(Duty_Cycle_Percent)  To adjust the duty cycle
p.change_frequency(frequency_Hz) To change the PWM frequency
p.pwm_polarity() To adjust the polarity, ie. changing duty cycle from percentage on to percentage off or vice versa
p.stop_pwm() To stop the PWM by setting the duty cycle to 0
p.pwm_close() To completely remove the PWM sysfs object.

For an OrangePi H3 board such as the PC + use PWM_pin = 0. As the PWM setup is different to the GPIO setup I didn't integrate the pin map present in this library. Its not an inconvenience however as this will likely be the same for all OPi H3 boards, but if you want to check what number you need to put in use:
"ls -l /sys/class/pwm/" This will show you all the PWMchips are on your system. it will be something like pwmchip0.
"echo 0 > /sys/class/pwm/pwmchip0/export"  this will create the PWM object. if an error returns saying no such device try 1 and so on. They are generally low numbers less than 5.
"ls /sys/class/pwm/pwmchip0" this will list all of the controls as well as the PWM object that was created.  The object is should  be called PWM0 (or PWM-whatever number worked)

The number following the PWM object name is the pin number you use in GPIO.PWM(PWM_pin, frequency_Hz, Duty_Cycle_Percent). It probably wouldn't be too difficult to make a list of pins for the various boards and use that as the pin mapping but I haven't.

I tried to leave a lot of comments as I was going along so hopefully the changes should make sense (this is my first time trying to contribute to someone's code so be nice 😉 ), I've added a demo to maybe showcase a bit better what is happening.

rm-hull commented on OPi/GPIO.py:
```diff
@@ -582,3 +582,104 @@ def cleanup(channel=None):
+        if channel in _exports: #check if there the pin is already assigned
+            raise RuntimeError("Channel {0} is already configured".format(channel))
+        
+        #if channel in _exports_pwm:
```
Remove commented out code

rm-hull commented on OPi/GPIO.py:
```diff
@@ -582,3 +582,104 @@ def cleanup(channel=None):
+            if e.errno == 16:   # Device or resource busy
+                if _gpio_warnings:
+                    warnings.warn("Channel {0} is already in use, continuing anyway. Use GPIO.setwarnings(False) to disable warnings.".format(channel), stacklevel=2)
+                #sysfs.unexport(pin) #leaving this here as a reminder to make a check. There will now be 2 types of pins that can be exported  
```
Remove commented out code

rm-hull reviewed (commented):
* Pretty sure this will fail the pep8 formatting, try running `tox -e qa` to see if there are errors
* There is a lot of code here, and no tests, please consider adding some
* Please add some documentation

rm-hull commented:
Thanks for submitting this, looks pretty useful - i added some comments inline

rm-hull commented:
In terms of documentation, the docs site (https://opi-gpio.readthedocs.io/en/latest/) is autogenerated from what is in `/docs` folder and code docstrings.

And tests, yes, have a study of the files in the `/tests` folder. You should be able to run them with `tox -e py37` for example (if you are python 3.7)

Fionnoch commented:
I've added the changes you suggested before. tox -e qa had an issue with the pi4B file but that was there before I started.

I've added some error checking tests, not a huge amount but it should cover the main errors a user could face.

Lastly I added a good bit to the documentation. Hopefully it should come out in the correct format on the webpage.

Fionnoch commented:
Hey do I have to resubmit a pull request to merge the changes I made or does this request reflect the most recent version?

rm-hull commented:
Sorry I haven’t had much chance to look at this again, thanks for the nudge. I will try and do that today

Fionnoch commented:
Hey just wanted to touch base, is there anything I can do help the review process?

rm-hull commented:
I pushed to master, but am struggling to deploy to pypi, it fails with:

```
error: Upload failed (400): Invalid value for blake2_256_digest. Error: Use a valid, hex-encoded, BLAKE2 message digest.
```

Fionnoch commented:
So I had a look and did a little bit of googling but couldn't figure out
what might be causing that, I'll setup a pypi account and try to
temporarily publish it via that to see if I can figure out whats causing
it. If I do manage to upload it I'll take it down right away.

Just letting you know so you don't think I'm republishing your own code!

On Sun, 25 Apr 2021 at 18:49, Richard Hull ***@***.***> wrote:

> I pushed to master, but am struggling to deploy to pypi, it fails with:
>
> error: Upload failed (400): Invalid value for blake2_256_digest. Error: Use a valid, hex-encoded, BLAKE2 message digest.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rm-hull/OPi.GPIO/pull/54#issuecomment-826362137>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APK2MXVWKHVWBR7VSG2HVULTKRIZHANCNFSM42AGBOMA>
> .
>

Fionnoch commented:
Hey so had a go uploading it to pypi and managed to get it uploaded to here
https://pypi.org/project/OPi.GPIOtest/0.5.0/ I didn't run into the same
issue, I ran it off an orange pi pc+.

I used
python setup.py sdist
twine upload dist/*
Getting twine installed on armbian was a bit of an issue (had to install
rust and  sudo apt-get install libffi-dev) and then it seemed to work. I
noticed that the documentation page didn't change, I assume thats separate
and controlled by yourself? Maybe that could be causing the issue. I added
info into the start of the OPi/GPIO.py file in the comments, I don't think
that it would cause an issue but I can't think of anything else that might
be messing up the upload.

Also I left the copy pypi file up so you can see but I'll take it down
whenever you'd like.

On Mon, 26 Apr 2021 at 09:31, Fionn O'Connell-Hussey <
***@***.***> wrote:

> So I had a look and did a little bit of googling but couldn't figure out
> what might be causing that, I'll setup a pypi account and try to
> temporarily publish it via that to see if I can figure out whats causing
> it. If I do manage to upload it I'll take it down right away.
>
> Just letting you know so you don't think I'm republishing your own code!
>
> On Sun, 25 Apr 2021 at 18:49, Richard Hull ***@***.***>
> wrote:
>
>> I pushed to master, but am struggling to deploy to pypi, it fails with:
>>
>> error: Upload failed (400): Invalid value for blake2_256_digest. Error: Use a valid, hex-encoded, BLAKE2 message digest.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rm-hull/OPi.GPIO/pull/54#issuecomment-826362137>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/APK2MXVWKHVWBR7VSG2HVULTKRIZHANCNFSM42AGBOMA>
>> .
>>
>

rm-hull commented:
ah, interesting ... i was using this script to upload: https://github.com/rm-hull/OPi.GPIO/blob/master/bin/make_release.sh

will try twine

Fionnoch commented:
Hopefully that works 🤞 I followed the guide at
***@***.***/how-to-upload-your-python-package-to-pypi-65edc5fe9c56
in case thats any help

On Wed, 28 Apr 2021 at 23:23, Richard Hull ***@***.***> wrote:

> ah, interesting ... i was using this script to upload:
> https://github.com/rm-hull/OPi.GPIO/blob/master/bin/make_release.sh
>
> will try twine
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rm-hull/OPi.GPIO/pull/54#issuecomment-828820349>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APK2MXQSHG5XZMTOKPXICQDTLCDFZANCNFSM42AGBOMA>
> .
>

Fionnoch commented:
Meant to ask did you ever get it to upload to pypi in the end? just curious
getting up a couple devices and was gonna pull from there if it was up, not
stressed if its not though.

On Thu, 29 Apr 2021 at 00:12, Fionn O'Connell-Hussey <
***@***.***> wrote:

> Hopefully that works 🤞 I followed the guide at
> ***@***.***/how-to-upload-your-python-package-to-pypi-65edc5fe9c56
> in case thats any help
>
> On Wed, 28 Apr 2021 at 23:23, Richard Hull ***@***.***>
> wrote:
>
>> ah, interesting ... i was using this script to upload:
>> https://github.com/rm-hull/OPi.GPIO/blob/master/bin/make_release.sh
>>
>> will try twine
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rm-hull/OPi.GPIO/pull/54#issuecomment-828820349>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/APK2MXQSHG5XZMTOKPXICQDTLCDFZANCNFSM42AGBOMA>
>> .
>>
>

vijayrajasekaran commented:
@Fionnoch Hello, thanks for this amazing contribution. I also own an Orange Pi PC Plus edition with Armbian Focal OS and I am unable to make this work.

I get the below error while trying to run the demo python example. Any workaround for this issue. Thanks.

`FileNotFoundError: [Errno 2] No such file or directory: '/sys/class/pwm/pwmchip0/export'`

The pwm directory is empty.

Fionnoch commented:
No worries at all thank you for the kind words! have you enabled PWM in the
armbian config? I've never used Arbian Focal but on the armbian version
I've used you needed to enable pwm with sudo armbian-config and I think its
under system > hardware, give that a go and let me know if you have any
luck.

On Wed, 29 Dec 2021 at 06:24, Vijay Rajasekaran ***@***.***>
wrote:

> @Fionnoch <https://github.com/Fionnoch> Hello, thanks for this amazing
> contribution. I also own an Orange Pi PC Plus edition with Armbian Focal OS
> and I am unable to make this work.
>
> I get the below error while trying to run the demo python example. Any
> workaround for this issue. Thanks.
>
> FileNotFoundError: [Errno 2] No such file or directory:
> '/sys/class/pwm/pwmchip0/export'
>
> The pwm directory is empty.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rm-hull/OPi.GPIO/pull/54#issuecomment-1002416676>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APK2MXQOPBNO7CG2RQVP3Q3UTKSRPANCNFSM42AGBOMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID: <rm-hull/OPi
> .***@***.***>
>

vijayrajasekaran commented:
> No worries at all thank you for the kind words! have you enabled PWM in the armbian config? I've never used Arbian Focal but on the armbian version I've used you needed to enable pwm with sudo armbian-config and I think its under system > hardware, give that a go and let me know if you have any luck.
> […](#)
> On Wed, 29 Dec 2021 at 06:24, Vijay Rajasekaran ***@***.***> wrote: @Fionnoch <https://github.com/Fionnoch> Hello, thanks for this amazing contribution. I also own an Orange Pi PC Plus edition with Armbian Focal OS and I am unable to make this work. I get the below error while trying to run the demo python example. Any workaround for this issue. Thanks. FileNotFoundError: [Errno 2] No such file or directory: '/sys/class/pwm/pwmchip0/export' The pwm directory is empty. — Reply to this email directly, view it on GitHub <[#54 (comment)](https://github.com/rm-hull/OPi.GPIO/pull/54#issuecomment-1002416676)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APK2MXQOPBNO7CG2RQVP3Q3UTKSRPANCNFSM42AGBOMA> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you were mentioned.Message ID: <rm-hull/OPi .***@***.***>

Just followed your instructions, enabled PWM and its working great. Thanks.
I am also able to use PA6 pin as PWM control using https://github.com/evergreen-it-dev/orangepwm with PyA20.gpio library installed using https://github.com/duxingkei33/orangepi_PC_gpio_pyH3

vijayrajasekaran commented:
Just followed your instructions, enabled PWM and its working great. Thanks.
I am also able to use PA6 pin as PWM control using https://github.com/evergreen-it-dev/orangepwm with PyA20.gpio library installed using https://github.com/duxingkei33/orangepi_PC_gpio_pyH3
